pipeline frame count and pcm range setting

juimonen opened the pull request:
ok, this is a little bit monster, but the patches are more or less related. Idea is to set pipeline component frame count from stream samplerate and scheduling time. Because of stream sample rate range is also set in the topology, there's patches to set that on top level m4. Because frame count also affects buffer size, there's also patches related to buffer re-size inside firmware. 50kHz topology is the only "non-related", but it is also good for testing this thing.

juimonen commented:
Here's some documentation related to this pr, if someone has time to read:
https://github.com/orgs/thesofproject/teams/sof-developers/discussions/5

lgirdwood commented on src/audio/host.c:
```diff
@@ -489,15 +490,31 @@ static int host_params(struct comp_dev *dev)
+#endif
+#if !defined CONFIG_DMA_GW
 	hd->period_bytes = dev->frames * comp_frame_bytes(dev);
+#endif
```
This is rounding up period size and not buffer size. It could also have impact on scheduling since we are processing one period per schedule time slice. 
Also the DW DMA has optimal burst size and alignment too like the DMA GW, so perhaps we need an additional ops to be added to the DMA driver to do this operation and simplify host.c

juimonen replied:
yes i was thinking a way to ask the dma alignment from the dma, but I still lack some knowledge about different dma's we have and their function. But yes exactly I would like to have method to get the alignment.

I guess the idea with the period_bytes alignment was, that if dma transfer _needs_ to be aligned I will always resize (round up) the host buffer (and period_bytes), so the host component will run with that scheduling, but rest of the pipeline will obey exactly the rate/sched. So for e.g. if you try to run pipeline 44 or 45 frames, host will push it up to next dma alignment (which probably in this case would be like running with 48 frames). Still rest of the pipeline will run with 44 as internally we don't have alignment restriction and host has ring buffer. As said I'm not sure how strictly the dma alignment should be obeyed...

lgirdwood commented on src/audio/host.c:
```diff
@@ -489,15 +490,31 @@ static int host_params(struct comp_dev *dev)
 	/* resize the buffer if space is available to align with period size */
 	buffer_size = hd->period_count * hd->period_bytes;
+	err = buffer_resize(hd->dma_buffer, buffer_size);
+
```
newline can be deleted.

lgirdwood commented on tools/topology/sof-apl-da7219.m4:
```diff
@@ -34,47 +34,53 @@ include(`platform/intel/dmic.m4')
 PIPELINE_PCM_ADD(sof/pipe-volume-playback.m4,
 	1, 0, 2, s32le,
-	48, 1000, 0, 0)
+	48, 1000, 0, 0,
```
So if frames come from srate (from runtime params) and scheduling time (from topology) surely we don't need the frame size here ? (and it can be deleted from macros)

juimonen replied:
removed in later commit and as suggested will merge it with another one.

lgirdwood commented on src/audio/pipeline.c:
```diff
@@ -59,9 +59,45 @@ struct op_data {
 
 static void pipeline_task(void *arg);
 
+static void set_frames(struct pipeline *p, struct comp_dev *current,
```
Would it not be better to walk the graph here from source to sink and have each component set it's number of frames (this would be based on rate, scheduling tick and source frames count).

juimonen replied:
sorry didn't get this. As I understand it, this is what is exactly happening when set_frames is called from op_downstream/op_upstream, which are walking through the graph from start to end...?

lgirdwood commented on tools/topology/m4/dai.m4:
```diff
@@ -178,7 +178,6 @@ define(`DAI_ADD',
 `undefine(`DAI_BUF')'
 `undefine(`DAI_PERIODS')'
 `undefine(`DAI_FORMAT')'
-`undefine(`SCHEDULE_FRAMES')'
```
Ok, this should be merged into previous patch, we also need to mark the token SCHEDULE_FRAMES as deprecated in FW and kernel headers (dont delete as it will be ignored now) and increment MINOR ABI.

juimonen replied:
well, with review of the last version Liam was asking for it :)

juimonen replied:
@xiulipan and yes the topology setting is removed, but I set it to always 0 on the actual pipeline struct

juimonen commented on src/audio/buffer.c:
```diff
@@ -94,6 +94,51 @@ struct comp_buffer *buffer_new(struct sof_ipc_buffer *desc)
+
+	if (!buffer) {
+		trace_buffer_error("resize error: can't to resize NULL buffer");
+		return -1;
```
ok, will do.

xiulipan commented on src/audio/src.c:
```diff
@@ -597,6 +597,8 @@ static struct comp_dev *src_new(struct sof_ipc_comp *comp)
 	cd->polyphase_func = src_polyphase_stage_cir;
 	src_polyphase_reset(&cd->src);
 
+	dev->output_rate = ipc_src->sink_rate;
```
I did not see any changes about SOF_TKN_SRC_RATE_OUT in your topology part.
Could you give some simple workflow about how this is set, just take the using sof-apl-src-50khz-pcm512x.tplg as exmple.

juimonen replied:
So the idea is, the component sets itself the output rate, if it _has_ fixed output rate. Like currently we have src and tone, that could use it. In other words it is not general feature that is set from topology for all components. As you noticed on other comment I rely on rzalloc to initialize it to 0 and components can set it after that (if they have fixed output rate).

xiulipan replied:
my point here is why mention 50khz in sof-apl-src-50khz-pcm512x.tplg.
I did not see any 50khz is src.
So you mean the actual output rate can comes from dai?
If it comes from dai, how could we do it only in downstream. The dai is the sink for src right?

juimonen replied:
oh yes, I see what you mean. Currently you have to match the dai and src. so 1) dai doesn't specify its samplerate, you have to deduce it from the parameters. 2) you can't specify src output rate from top level topology, you should do that in pipe-src-playback, where it is currently hardcoded to 48kHz.

xiulipan replied:
so could you help me understand how sof-apl-src-50khz-pcm512x.tplg pass the 50kzh rate into the src part `dev->output_rate`

I think this is needed for the set_frame?

xiulipan commented on src/include/sof/audio/component.h:
```diff
@@ -211,6 +211,7 @@ struct comp_dev {
 	uint64_t position;		/* component rendering position */
 	uint32_t frames;		/* number of frames we copy to sink */
 	uint32_t frame_bytes;		/* frames size copied to sink in bytes */
+	uint32_t output_rate;		/* 0 means all output rates are fine */
```
I see some set to 0 in `buffer_new`, but no for this.
```c
        buffer->avail = 0;
	buffer->connected = 0;
```
But it seems we used rzalloc for all comp.

juimonen replied:
yes, see previous comment.

juimonen commented on src/audio/buffer.c:
```diff
@@ -94,6 +94,51 @@ struct comp_buffer *buffer_new(struct sof_ipc_buffer *desc)
+
+	buffer_zero(buffer);
+
+	return 0;
```
sure, I'll look into it.

juimonen replied:
@xiulipan Thanks for the review! I'll try to answer your questions:
1) The idea is that new method set_frames() is called with pipeline params, that is walking through the whole graph. So host is getting frames from stream rate / sched (aligned with dma burst), next component will get frames from previous (if not fixed output_rate, when component will set the frames itself from its own rate) and so on...
2) yes topology changes are big, I could have left the old params there and just not use them, but 
@lgirdwood asked for removal and I myself agree it would be confusing to leave them there. And as you noticed the frame count is still in the pipeline struct that is parsed in kernel (I just set it to 0 and it is not used anywhere).

juimonen commented on src/audio/host.c:
```diff
@@ -489,15 +490,31 @@ static int host_params(struct comp_dev *dev)
+	hd->period_bytes = ceil_divide(dev->frames * comp_frame_bytes(dev),
+				       PLATFORM_HDA_BUFFER_ALIGNMENT) *
+		PLATFORM_HDA_BUFFER_ALIGNMENT;
+#endif
```
Sure. The dma related #ifdefs look quite ugly in host.c and I'm not totally confident I understand all possible cases. As Liam suggested we should have ops in dma to ask for optimal burst size.

lgirdwood replied:
@juimonen Do you have the DMA ops ready yet, I dont see them in the latest push ?

juimonen commented on tools/topology/sof-apl-src-50khz-pcm512x.m4:
```diff
@@ -0,0 +1,72 @@
+dnl     frames, deadline, priority, core)
+
+# Playback pipeline 1 on PCM 0 using max 2 channels of s24le.
+# Schedule 48 frames per 1000us deadline on core 0 with priority 0
```
oh yes, good point. will do.

juimonen commented on tools/topology/sof-apl-src-50khz-pcm512x.m4:
```diff
@@ -0,0 +1,72 @@
+
+#SSP 5 (ID: 0)
+DAI_CONFIG(SSP, 5, 0, SSP5-Codec,
+	SSP_CONFIG(I2S, SSP_CLOCK(mclk, 19200000, codec_mclk_in),
```
@xiulipan I was running this topology with up2 and hifiberry with sound...?

juimonen replied:
I got the ssp params from @singalsu, maybe you could comment here also...

singalsu replied:
The used I2S format is such that MCLK is 19.2 MHz and bit clock is 3.2 MHz (19.2/6). With 64 or 32*2 bits frame that gives the 50 kHz sample rate.

juimonen commented on src/audio/pipeline.c:
```diff
@@ -111,7 +145,6 @@ static void connect_downstream(struct pipeline *p, struct comp_dev *start,
 
 	/* complete component init */
 	current->pipeline = p;
-	current->frames = p->ipc_pipe.frames_per_sched;
```
ok, will look into it. I didn't think the testbench at all for this.

juimonen replied:
ok, look into testbench, I don't see that fs_in or fs_out is actually used anywhere? with this patch If it is 0 it is set to 0 (again, because pipeline frames is always 0), if it is set from command line it is not touched. @ranj063 any thoughts?

juimonen commented on src/audio/buffer.c:
```diff
@@ -94,6 +94,51 @@ struct comp_buffer *buffer_new(struct sof_ipc_buffer *desc)
+		return -1;
+	}
+
+	desc->size = size;
```
@xiulipan sorry didin't get this, could you elaborate a bit more?

juimonen replied:
ok, got it. will try it out.

juimonen commented:
@lgirdwood @xiulipan thanks for the review. I updated the PR. I tried to incorporate the suggestions, but there are still couple of comments I did not totally understand. 

This PR should be really tested by QA with all possible platforms and topologies as I'm limited to up2...

markyang commented:
**The test results of PR842 are as follows:**

DUT | Device Codec | Driver loaded | PCM devices list | HDMI/DP playback | Analogy playback | Analogy capture | DMIC capture | Volume Change | Unload/reload   (5 iterations) | Unload/reload   (100 iterations) | Runtime PM | S3/S0ix w/o playback   (50 iterations) | S3/S0ix w/    I2C playback   (50 iterations) | S3/S0ix w/    HDMI playback   (50 iterations)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
BYT Minnow   Board | rt5651 | Passed | Passed | N/A | Passed | Passed | N/A | Passed | Passed | Passed | N/A | N/A | N/A | N/A
APL UP2 | pcm512x | Passed | Passed | Passed | Passed | N/A | N/A | Passed | Passed | Passed | Passed | Passed \| 50 | Passed \| 50 | Failed \| 1 (#788)   Known issue


**Test env:**
sof master: 03ababfb (pr842 was applied)
kernel sof-dev: a5dd3186 (latest)
tplg: sof-apl-pcm512x.tplg (pr842 was applied)
tplg: sof-byt-rt5651.tplg (pr842 was applied)
tplg: test-ssp0-mclk-0-I2S-volume-s16le-s16le-48k-24000k-codec.tplg (latest)

lgirdwood commented:
@markyang great results :)
@juimonen can you check CI build failure, you may need to restart build if results are older than 1 day.

juimonen commented:
@lgirdwood I'll rebase, however it goes to tomorrow, there was tomasz' patch for pipeline refactoring, which I have to resolve...

lgirdwood commented:
@juimonen @tlauda lets get Tomeks PR merged first then we can do this on top.

juimonen commented:
@lgirdwood, rebased and updated. The frame setting code now changed, because of re-factored pipeline code. I did test basic src topologies and they seem to work. I tried to test also 8-192khz range to 48kHz dai, but there's some problem. I think the kernel sof/pcm.c is currently tied to dai max rate, which is not going to work in the case mentioned before. I don't know however is it a problem for this pr now...

juimonen commented:
this PR is failing in cmocka test: test_audio_pipeline_complete_connect_downstream_full
which is testing frame setting from pipeline. Well this PR is removing setting frames from pipeline and instead using stream parameters to calculate it. So is the cmocka test invalid for this PR or I'm not understanding what is happening? @tlauda can you point me to right person writing the tests? I saw your name in the file...

lgirdwood commented on src/audio/pipeline.c:
```diff
@@ -258,6 +257,8 @@ int pipeline_free(struct pipeline *p)
 {
 	struct pipeline_data *ppl_data = data;
+	int dl = current->pipeline->ipc_pipe.deadline;
+	int dl_m = current->pipeline->ipc_pipe.deadline - 1;
```
care to expand the variable names into something more meaningful.

juimonen replied:
removed the variables and add comment

lgirdwood commented:
@juimonen just a minor comment and looks like a UT failure.

juimonen commented:
@lgirdwood I fixed the unit test, I hope the fix was ok. I'm still seeing strange build errors, some time outs and klockwork "indexing" errors...

zrombel commented:
On QuickBuild in python regression test for this PR we are encountering DSP panic when sending ipc_stream_pcm_params to FW. This problem applies to APL, CNL and ICL builds. 
PR_REGRESSION: https://quickbuild.igk.intel.com/build/3457197

IPC log (after creating topology when trying to create stream):

Creating stream (48000Hz, 2, 32b/32b).
Sending StreamPmcParams (60010000, 00000000)
Payload:
0000006C 60010000 00000000 00000000
00000000 00000000 00000054 0000001C
00000000 00000000 00000000 00000000
00000000 00000000 00000001 00000002
00000000 0000BB80 00020001 00040004
00003F00 00000000 00000000 00000000
00000000 00000000 00000000

Notification Ipc (8DEA0000, 00040800)
Payload:
0000006C 70000000 00000000 00000000
00000000 00000000 0000003C 00010001
00000000 00000000 00000000 00000000
00000000 00000000 66330000 00323733
03000000 00000000 00000000 00000000
00000000 00000000 00000000 00000000
00000000 00000000 00000000

Full IPC log can be found on QuickBuild or I can provide it. All builds was made with xtensa-smp architecture. I will try to run test on APL with xtensa-up to see if problem is architecture dependent.

lgirdwood commented on tools/topology/sof-apl-src-50khz-pcm512x.m4:
```diff
@@ -0,0 +1,71 @@
+dnl     frames, deadline, priority, core)
+
+# playback DAI is SSP5 using 2 periods
+# Buffers use s24le format, with 48 frame per 1000us on core 0 with priority 0
```
Does this comment need aligned for 50kHz rate ?

lgirdwood commented:
@juimonen can you add dma_get_attribute(dma, uint32_t type, uint32_t *value) as a DMA driver op.

juimonen commented:
@lgirdwood added dma ops. is this something you we're thinking?

lgirdwood commented:
@juimonen can you check CI

juimonen commented:
@lgirdwood as I understand somehow the klockwork run failed (no connection etc.)... I don't know how to resolve it.

Anyway, we should probably ask another QA run for this as many things around pipeline have changed. I know it takes time, but I'm not too confident otherwise.

lgirdwood commented:
@juimonen CI can be restarted by doing a force push.

mluckix commented:
What is the status of this PR? Will it be continued or rejected?

juimonen commented:
@mluckix, @tlauda is working on quite big pipeline refactoring (don't remember the pr), it is clashing with this quite heavily. My thinking was, that after that is done/merged I can come back to this and rebase, and see is there anything to be done or is the functionality in tomasz' pr.

tlauda commented:
@juimonen My last big PR for pipeline is #1045. After that I expect only minor changes to the pipeline and more major ones to the schedule API. I think after #1045 get merged you are good to go.

mluckix commented:
@juimonen, @tlauda big thanks for the status. I have two minor bugs which will be resolved by this PR (I have just checked it). We planed to return error when host's runtime params result in buffer size that do not fit DMA chunk alignment. In such cases @juimonen's PR properly resizes the buffer.

lgirdwood commented:
@juimonen can you fix the conflicts for QA and @mluckix.

tlauda commented:
@juimonen Could you update and rebase?

juimonen commented:
@tlauda arggh, quite nasty rebase ahead, will take some time and testing...

juimonen commented:
@tlauda updated. I tested only basic and 50kHz pipe in up2. So needs heavy testing and QA rounds. @keyonjie can you check the kwd topologies as I was not sure how to modify those...

tlauda commented on src/audio/buffer.c:
```diff
@@ -80,21 +80,55 @@ struct comp_buffer *buffer_new(struct sof_ipc_buffer *desc)
+	if (buffer->alloc_size == size)
+		return 0;
+
+	if (buffer->addr)
```
We should lock this part to block the access to the buffer during resizing.

tlauda commented on src/audio/buffer.c:
```diff
@@ -80,21 +80,55 @@ struct comp_buffer *buffer_new(struct sof_ipc_buffer *desc)
+		return -EINVAL;
+	}
+
+	buffer->addr = rballoc(RZONE_RUNTIME, desc->caps, size);
```
Could we use `RZONE_BUFFER` to be consistent?
I know it doesn't matter, but still.

tlauda commented:
@juimonen I know you are currently involved in different topics, but could we wrap up this one? Couple of comments from my side, missing DMA alignment commits and conflicts to be resolved. It shouldn't take much time and @lgirdwood has already approved.

juimonen commented:
@tlauda sorry for the delay, I'll try to get it together on Friday...

juimonen commented:
@tlauda updated. But we really should run some extended test for this.... this is touching everywhere.
@mengdonglin can we run some testing for this?

tlauda commented on tools/topology/platform/intel/intel-generic-dmic.m4:
```diff
@@ -20,13 +20,13 @@ dnl     frames, deadline, priority, core)
 # Schedule 48 frames per 1000us deadline on core 0 with priority 0
```
@juimonen Comments need to be adjusted in every topology.

juimonen replied:
ok

tlauda commented on src/audio/host.c:
```diff
@@ -532,11 +533,17 @@ static int host_params(struct comp_dev *dev)
 #if CONFIG_DMA_GW
 		buffer_count = hd->period_count;
-		buffer_single_size = dev->frames * comp_frame_bytes(dev);
+		buffer_single_size = ceil_divide(dev->frames *
```
We can use ALIGN_UP macro.

juimonen replied:
ok

tlauda commented on src/audio/host.c:
```diff
@@ -532,11 +533,17 @@ static int host_params(struct comp_dev *dev)
 		buffer_count = 1;
-		buffer_single_size = hd->period_count * dev->frames *
-			comp_frame_bytes(dev);
+		buffer_single_size = ceil_divide(hd->period_count *
```
We can use ALIGN_UP macro.

juimonen replied:
ok

tlauda commented on src/audio/host.c:
```diff
@@ -552,7 +559,10 @@ static int host_params(struct comp_dev *dev)
 		hd->period_count = cconfig->periods_source;
 		buffer_count = hd->period_count;
-		buffer_single_size = dev->frames * comp_frame_bytes(dev);
+		buffer_single_size = ceil_divide(dev->frames *
```
We can use ALIGN_UP macro.

juimonen replied:
ok

tlauda commented on src/audio/host.c:
```diff
@@ -566,15 +576,30 @@ static int host_params(struct comp_dev *dev)
 
+#if defined CONFIG_DMA_GW
 	/* calculate period size based on config */
+	hd->period_bytes = ceil_divide(dev->frames * comp_frame_bytes(dev),
```
We can use ALIGN_UP macro.

juimonen replied:
ok

tlauda commented on src/audio/pipeline.c:
```diff
@@ -305,6 +304,17 @@ static int pipeline_comp_params(struct comp_dev *current, void *data, int dir)
 	/* send current params to the component */
 	current->params = ppl_data->params->params;
 
+	/* set frames from samplerate/period, but round integer up */
```
We can use ALIGN_UP macro.

juimonen replied:
ok

juimonen replied:
hmm so I want the rate to be divisible with frames, but change the frames (not rate)... can we still use align up?

tlauda replied:
@juimonen It can stay as it is.

tlauda commented on test/cmocka/src/audio/pipeline/pipeline_connect_upstream.c:
```diff
@@ -108,11 +108,8 @@ static void test_audio_pipeline_complete_connect_downstream_variable_set
-	result.sched_comp->frames, test_data->p.ipc_pipe.frames_per_sched
-	);
 	assert_ptr_equal(result.sched_comp->pipeline, &result);
+	assert_ptr_equal(test_data->first->pipeline, &result);
```
Typo in commit description: `successful`.

juimonen replied:
ok

tlauda commented on src/include/sof/dma.h:
```diff
@@ -168,6 +171,8 @@ struct dma_ops {
 	int (*get_data_size)(struct dma *dma, int channel, uint32_t *avail,
 			     uint32_t *free);
+
+	int (*get_attribute)(struct dma *dma, uint32_t type, uint32_t *value);
```
Please split this commit into two separate:
1. Adding ops to dma.
2. Using ops in host.

juimonen replied:
ok

tlauda commented on src/drivers/dw/dma.c:
```diff
@@ -1322,6 +1322,22 @@ static int dw_dma_get_data_size(struct dma *dma, int channel, uint32_t *avail,
 	return 0;
 }
 
+static int dw_get_attribute(struct dma *dma, uint32_t type,
```
Please align name with the rest of the functions: `dw_dma_get_attribute`.

tlauda commented on src/drivers/dw/dma.c:
```diff
@@ -1322,6 +1322,22 @@ static int dw_dma_get_data_size(struct dma *dma, int channel, uint32_t *avail,
+		*value = DW_DMA_BUFFER_ALIGNMENT;
+		break;
+	default:
+		ret = -EINVAL;
```
It's good practice to add break also after default label.

juimonen replied:
ok

tlauda commented on src/drivers/intel/cavs/hda-dma.c:
```diff
@@ -962,6 +962,22 @@ static int hda_dma_data_size(struct dma *dma, int channel, uint32_t *avail,
+		*value = PLATFORM_HDA_BUFFER_ALIGNMENT;
+		break;
+	default:
+		ret = -EINVAL;
```
It's good practice to add break also after default label.

juimonen replied:
ok

tlauda commented on src/include/sof/dw-dma.h:
```diff
@@ -153,6 +153,9 @@
 #define trace_dwdma_error(__e, ...) \
 	trace_error(TRACE_CLASS_DMA, __e, ##__VA_ARGS__)
 
+#define DW_DMA_MAX_NR_CHANNELS	8
```
Not used anywhere.

juimonen replied:
ok

tlauda commented:
@juimonen Just resolve conflicts and my review suggestions and we can give it a try during regression testing.

juimonen commented:
@tlauda updated, I'm, not sure I can use align-up in pipeline frames calculation and I can't access the pr-tools-build failure... server down etc?

tlauda commented:
@xiulipan Can you take a look?

lgirdwood commented on src/audio/pipeline.c:
```diff
@@ -281,6 +280,17 @@ static int pipeline_comp_params(struct comp_dev *current, void *data, int dir)
 	/* send current params to the component */
 	current->params = ppl_data->params->params;
 
+	/* set frames from samplerate/period, but round integer up */
```
Can we roll this into https://github.com/thesofproject/sof/issues/1280 and  https://github.com/orgs/thesofproject/teams/sof-developers/discussions/9 . These are related and should be in the same PR for QA.

juimonen commented on tools/topology/sof-apl-src-50khz-pcm512x.m4:
```diff
@@ -0,0 +1,72 @@
+#
+# Topology for Apollolake UP^2 with pcm512x codec with sample rate
```
@lgirdwood I think this was request from @plbossart at some point... the reason it is in this PR is that this PR makes it work, it was not possible to run it before these changes. So it is here for both "feature request" and for testing. However, It can be made as another PR or dropped if not needed anymore.

juimonen commented:
@lgirdwood @tlauda so I can chop this to different PR's just tell me what would you like to have...
so for example:
PR 1) add dma ops, add buffer resize, add using 2 previous in host
PR 2) "optional" 50khz topology
PR 3) pipeline frame count, related topology changes, cmocka fix

@lgirdwood I'm not sure I should separate in PR3 the functional and topology changes? I can do that if you really want. They don't too much make change as separate items...

juimonen commented on src/audio/buffer.c:
```diff
@@ -56,21 +56,64 @@ struct comp_buffer *buffer_new(struct sof_ipc_buffer *desc)
 	return buffer;
 }
 
+int buffer_resize(struct comp_buffer *buffer, uint32_t size)
```
@lgirdwood hmm I might be here a little bit out of my expert zone, but sure can try.

juimonen replied:
@lgirdwood @tlauda so something like this?
https://github.com/thesofproject/sof/pull/1533

lgirdwood commented:
@juimonen yep you can merge PR2 and PR3.

xiulipan commented:
@tlauda @juimonen 

01.org sync have issue with proxy.

Please check the tplg build 

> ALSA lib parser.c:219:(tplg_parse_config) error: unknown section )
> 
> ALSA lib parser.c:219:(tplg_parse_config) error: unknown section pipe-sof-apl-keyword-detect
> 
> m4: test-capture.m4: 45: fatal error: Invalid parameters (13) to PIPELINE_PCM_DAI_ADD

juimonen commented:
@xiulipan hmm my local build for topologies is working. I'm using the same docker setup as CI. I now force pushed the same stuff with commits reordered... let's see what happens

xiulipan commented:
@juimonen 
The PR code works fine on my local too. 
Let me check why Jenkins CI will fail.

xiulipan commented:
@juimonen 
The jenkins CI test with `./scripts/build-tools.sh -t`, thus we will also check the test topology build. That will be a lot of tplg and some of them is failed due to you have changed the basic m4 code.

juimonen commented:
@xiulipan aah ok now I get it, thanks, I'll see the test build also

lgirdwood commented:
@xiulipan @juimonen can you confirm if CI failure is false positive ?

xiulipan commented:
@lgirdwood @juimonen 
there must be some error in BYT_NOCODEC
https://sof-ci.01.org/sofpr/PR842/build2206/devicetest/BYT_MB_NOCODEC/verify_pcm_list.sh/slogger.txt
It seems more memory space is needed for this patch.

```bash
    0      2       BUFFER          14420443.645833         3.541667   src/audio/buffer.c:29  	buffer_new()
    0      1          MEM          14420450.364583         6.718750      src/lib/alloc.c:246 	error: 42 blocks needed for allocation but only 41 blocks are remaining
    0      1       BUFFER          14420460.885417        10.520833   src/audio/buffer.c:52  	buffer_new() error: could not alloc size = 16000 bytes of type = 65
    0      1          IPC          14420470.260417         9.375000        src/ipc/ipc.c:193 	ipc_buffer_new() error: buffer_new() failed
    0      1          IPC          14420477.083333         6.822917    src/ipc/handler.c:923 	ipc: pipe 3 buffer 23 creation failed -12
```

For APL_PCM512X device, it is just timeout. I checked the result after timeout. No error.

@juimonen 
Please check and fix BYT nocodec issue.

juimonen commented:
@xiulipan I saw the same strange 16000 bytes buffer in some of my local conf files, but after recompilation they are gone.... and after force pushing the checks pass. So I'm not sure what was causing that... makes me suspicious we have some parallel build issue etc.

lgirdwood commented:
@juimonen can you provide more details, is this a topology build issue or FW issue. Could be we have some invalid dependencies somewhere that is causing stale data to be used. @jajanusz fyi.

juimonen commented:
@xiulipan sorry to bother you again, any idea what is the current CI error?

juimonen commented:
@lgirdwood it was topology build issue... I haven't seen it anymore, my assumption is that I did not fix anything related to it, but it went away... doesn't rule out that I had some erroneous force push in between etc.

lgirdwood commented:
@juimonen Can you confirm you are using #1533 now for resize ? Not seeing this in the changes.

juimonen commented:
@lgirdwood not yet. I was hoping to get this clear without errors... I still see some CI code check error, which I can't interpret. I'll modify to this to use realloc and them possible division to separate PR. I would hope 1 extra manual testing round with this when CI shows green light.

juimonen commented:
@xiulipan any idea about the unit test fail? I'm having trouble of interpreting it...

xiulipan commented:
@tlauda @zrombel @jajanusz Any idea about QB unit test here
@juimonen Please ask QB issues with above guys

jajanusz commented:
@juimonen made fix for you https://github.com/juimonen/sof/pull/1 however you should just squash commits that update unit tests with commits that add api-breaking changes, because otherwise you'll break `git bisect` for unit tests

juimonen commented:
@jajanusz thanks!

juimonen commented:
hmm ok. people are cherry picking commits from this PR and they get merged... I have to make separate PR's from rest of this, it starts to be really messy.

juimonen commented:
made https://github.com/thesofproject/sof/pull/160 for host buffer resize

juimonen commented:
and removed the stuff in https://github.com/thesofproject/sof/pull/1603 from this PR...

lgirdwood commented:
@juimonen ok great, I think it now just needs the conflicts fixed now.

juimonen commented:
@lgirdwood @mengdonglin I'd still like to have QA test round as this is very invasive... I can't test most platforms on my own.

tlauda commented:
@juimonen Can we get this rebased and maybe finally merged?

juimonen commented:
@tlauda rebased. I think @mengdonglin is on vacation this week and I would like to have the extended testing for this... This is touching all topology files and I can't test them myself.

tlauda commented:
@lgirdwood @xiulipan Please rereview ASAP to avoid another rebase.

dbaluta commented on tools/topology/CMakeLists.txt:
```diff
@@ -58,6 +58,7 @@ set(TPLGS
 	"sof-apl-dmic-a2ch-b4ch\;sof-apl-dmic-a2ch-b4ch"
 	"sof-apl-dmic\;sof-apl-dmic-4ch\;-DCHANNELS=4"
```
@juimonen Please add ":" in commit message after topology. E.g: topology: add ...

juimonen replied:
ok, added

xiulipan commented on tools/topology/sof/pipe-src-volume-playback.m4:
```diff
@@ -61,13 +61,13 @@ W_PGA(0, PIPELINE_FORMAT, 2, 2, playback_pga_conf, LIST(`		', "PIPELINE_ID Maste
 
 # Playback Buffers
 W_BUFFER(0, COMP_BUFFER_SIZE(3,
-	COMP_SAMPLE_SIZE(PIPELINE_FORMAT), PIPELINE_CHANNELS, SCHEDULE_FRAMES),
```
Do we miss some setting to 
`W_VENDORTUPLES(media_src_tokens, sof_src_tokens, LIST(`		', `SOF_TKN_SRC_RATE_OUT	"PIPELINE_RATE"'))`
in this file?

juimonen replied:
yes, you are right. As this PR has been lingering for almost a year, topologies come an go...

xiulipan commented on tools/topology/sof-apl-src-50khz-pcm512x.m4:
```diff
@@ -0,0 +1,72 @@
+
+dnl PIPELINE_PCM_ADD(pipeline,
+dnl     pipe id, pcm, max channels, format,
+dnl     frames, deadline, priority, core)
```
```
dnl PIPELINE_PCM_ADD(pipeline,
dnl     pipe id, pcm, max channels, format,
dnl     period, priority, core,
dnl     pcm_min_rate, pcm_max_rate, pipeline_rate,
dnl     time_domain, sched_comp)```

juimonen replied:
ok

xiulipan commented on tools/topology/sof-apl-src-50khz-pcm512x.m4:
```diff
@@ -0,0 +1,72 @@
+dnl     frames, deadline, priority, core)
+
+# Playback pipeline 1 on PCM 0 using max 2 channels of s24le.
+# Schedule 48 frames per 1000us deadline on core 0 with priority 0
```
Do we also need to change the comments here?
PS: I think this commit should go after the `topology: enable pcm range and pipeline rate and remove frame count`

juimonen replied:
ok

juimonen commented:
@xiulipan it could/should be used with any/all pipelines. The old "frames" is going away, because we can calculate it automatically from samplerate&scheduletime. Of course difference on these new rates doesn't make much sense with pipeline without SRC.

juimonen commented:
@xiulipan pushed new version. I tried to address your comments.

tlauda commented:
SOFCI TEST

tlauda commented:
@juimonen Do we need any additional testing or you think CI is enough? I would like to merge it as soon as possible and avoid another rebase.

juimonen commented:
@tlauda I'm not sure if all use cases with kwd and more exotic things work as intended... I'm not sure is anyone testing those. Anyway, it seems basic stuff is working and I'm fine with reverting or fixing if things arise.

tlauda commented:
@juimonen CI passed (BYT machine with codec is currently down).
